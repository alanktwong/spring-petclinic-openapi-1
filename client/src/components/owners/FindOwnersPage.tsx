import * as React from 'react';

import { Link, RouteComponentProps, withRouter } from 'react-router-dom';
import { Owner, OwnerApi } from 'petclinic-api';

import OwnersTable from './OwnersTable';

import * as H from 'history';

import QueryString from 'query-string';

import { ErrorMessage, Field, Form, Formik } from 'formik';

interface IFindOwnersPageProps extends RouteComponentProps {}

interface IFindOwnersPageState {
  owners: Owner[];
}

const getFilterFromLocation = (location: H.Location): string | undefined => {
  const query = QueryString.parse(location.search);
  const lastName = query['lastName'];
  if (Array.isArray(lastName)) {
    return lastName[0];
  }
  if (lastName === null) {
    return undefined;
  }
  return lastName;
};

class FindOwnersPage extends React.Component<
  IFindOwnersPageProps,
  IFindOwnersPageState
> {
  constructor(props) {
    super(props);

    this.state = { owners: [] };
  }

  async componentDidMount() {
    const filter = getFilterFromLocation(this.props.location);
    if (typeof filter === 'string') {
      // only load data on mount (initially) if filter is specified
      // i.e. lastName query param in URI was set
      this.fetchData(filter);
    }
  }

  /**
   * Invoked when the submit button was pressed.
   *
   * This method updates the URL with the entered lastName. The change of the URL
   * leads to new properties and thus results in rendering.
   */
  submitSearchForm = values => {
    const { filter } = values;

    this.props.history.push({
      pathname: '/owners/list',
      search: `?lastName=${encodeURIComponent(filter || '')}`,
    });

    this.fetchData(filter);
  };

  /**
   * Actually loads data from the server
   */
  async fetchData(filter: string) {
    const owners = await new OwnerApi().listOwners({ lastName: filter || '' });

    this.setState({ owners });
  }

  render() {
    const filter = getFilterFromLocation(this.props.location) || '';
    const owners = this.state.owners;

    const showResults: boolean = filter.length > 0 || owners.length > 0;

    return (
      <div>
        <section>
          <h2>Find Owners</h2>
          <Formik initialValues={{ filter }} onSubmit={this.submitSearchForm}>
            {() => (
              <Form className="form-horizontal">
                <div className="form-group">
                  <div className="control-group" id="lastName">
                    <label className="col-sm-2 control-label">Last name </label>
                    <div className="col-sm-10">
                      <Field
                        name="filter"
                        className="form-control"
                        size={30}
                        maxLength={80}
                      />
                      <ErrorMessage name="filter" component="div" />
                    </div>
                  </div>
                </div>
                <div className="form-group">
                  <div className="col-sm-offset-2 col-sm-10">
                    <button type="submit" className="btn btn-default">
                      Find Owner
                    </button>
                  </div>
                </div>
              </Form>
            )}
          </Formik>
        </section>
        {showResults && <OwnersTable owners={owners} />}
        <Link className="btn btn-default" to="/owners/new">
          {' '}
          Add Owner
        </Link>
      </div>
    );
  }
}

export default withRouter(FindOwnersPage);
