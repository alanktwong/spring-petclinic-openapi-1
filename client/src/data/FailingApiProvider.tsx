import * as React from 'react';
import { FailingApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithFailingApiProps {
  failingApi: FailingApi;
}

export const FailingApiContext = React.createContext(new FailingApi());

export const withFailingApi = <P extends WithFailingApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithFailingApi extends React.Component<
    Subtract<P, WithFailingApiProps>
  > {
    static contextType = FailingApiContext;
    context!: React.ContextType<typeof FailingApiContext>;

    render() {
      return <Component failingApi={this.context} {...(this.props as P)} />;
    }
  };
