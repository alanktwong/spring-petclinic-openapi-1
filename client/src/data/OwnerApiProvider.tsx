import * as React from 'react';
import { OwnerApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithOwnerApiProps {
  ownerApi: OwnerApi;
}

export const OwnerApiContext = React.createContext(new OwnerApi());

export const withOwnerApi = <P extends WithOwnerApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithOwnerApi extends React.Component<Subtract<P, WithOwnerApiProps>> {
    static contextType = OwnerApiContext;
    context!: React.ContextType<typeof OwnerApiContext>;

    render() {
      return <Component ownerApi={this.context} {...(this.props as P)} />;
    }
  };
