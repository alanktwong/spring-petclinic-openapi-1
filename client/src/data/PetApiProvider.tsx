import * as React from 'react';
import { PetApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithPetApiProps {
  petApi: PetApi;
}

export const PetApiContext = React.createContext(new PetApi());

export const withPetApi = <P extends WithPetApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithPetApi extends React.Component<Subtract<P, WithPetApiProps>> {
    static contextType = PetApiContext;
    context!: React.ContextType<typeof PetApiContext>;

    render() {
      return <Component petApi={this.context} {...(this.props as P)} />;
    }
  };
