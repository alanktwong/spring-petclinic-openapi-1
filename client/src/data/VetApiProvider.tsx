import * as React from 'react';
import { VetApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithVetApiProps {
  vetApi: VetApi;
}

export const VetApiContext = React.createContext(new VetApi());

export const withVetApi = <P extends WithVetApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithVetApi extends React.Component<Subtract<P, WithVetApiProps>> {
    static contextType = VetApiContext;
    context!: React.ContextType<typeof VetApiContext>;

    render() {
      return <Component vetApi={this.context} {...(this.props as P)} />;
    }
  };
