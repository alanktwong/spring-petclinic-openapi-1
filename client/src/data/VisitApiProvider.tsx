import * as React from 'react';
import { VisitApi } from 'petclinic-api';
import { Subtract } from 'utility-types';

export interface WithVisitApiProps {
  visitApi: VisitApi;
}

export const VisitApiContext = React.createContext(new VisitApi());

export const withVisitApi = <P extends WithVisitApiProps>(
  Component: React.ComponentType<P>
) =>
  class WithVisitApi extends React.Component<Subtract<P, WithVisitApiProps>> {
    static contextType = VisitApiContext;
    context!: React.ContextType<typeof VisitApiContext>;

    render() {
      return <Component visitApi={this.context} {...(this.props as P)} />;
    }
  };
