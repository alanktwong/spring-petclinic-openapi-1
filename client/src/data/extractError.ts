import { IError } from '../types';

const extractError = (response): Promise<IError> => {
  if (response.json) {
    return response.json();
  } else {
    return Promise.resolve({ message: `${response}` });
  }
};

export default extractError;
