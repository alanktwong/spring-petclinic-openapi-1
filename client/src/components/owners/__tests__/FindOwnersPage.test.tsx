import React from 'react';
import renderer from 'react-test-renderer';
import FindOwnersPage from '../FindOwnersPage';
import waitForExpect from 'wait-for-expect';
import { OwnerApiContext } from '../../../data/OwnerApiProvider';
import { OwnerApi } from 'petclinic-api';
import { StaticRouter } from 'react-router-dom';

it('renders correctly', async () => {
  const mock = new OwnerApi();
  mock.listOwners = async () => [
    {
      id: 1,
      firstName: 'firstName1',
      lastName: 'lastName1',
      address: 'address1',
      city: 'city1',
      telephone: '012345',
      pets: [],
    },
  ];

  const component = renderer.create(
    <StaticRouter location="/owners/list?lastName=lastName1">
      <OwnerApiContext.Provider value={mock}>
        <FindOwnersPage />
      </OwnerApiContext.Provider>
    </StaticRouter>
  );

  const page = component.root.find(node => node.type.name === 'FindOwnersPage');

  await waitForExpect(() => {
    expect(page.instance.state.owners[0].id).toBe(1);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
