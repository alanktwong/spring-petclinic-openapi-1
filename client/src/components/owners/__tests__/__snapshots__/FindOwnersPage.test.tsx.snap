// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <section>
    <h2>
      Find Owners
    </h2>
    <form
      className="form-horizontal"
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <div
          className="control-group"
          id="lastName"
        >
          <label
            className="col-sm-2 control-label"
          >
            Last name 
          </label>
          <div
            className="col-sm-10"
          >
            <input
              className="form-control"
              maxLength={80}
              name="filter"
              onBlur={[Function]}
              onChange={[Function]}
              size={30}
              value="lastName1"
            />
          </div>
        </div>
      </div>
      <div
        className="form-group"
      >
        <div
          className="col-sm-offset-2 col-sm-10"
        >
          <button
            className="btn btn-default"
            type="submit"
          >
            Find Owner
          </button>
        </div>
      </div>
    </form>
  </section>
  <section>
    <h2>
      1
       Owners found
    </h2>
    <table
      className="table table-striped"
    >
      <thead>
        <tr>
          <th>
            Name
          </th>
          <th
            className="hidden-sm hidden-xs"
          >
            Address
          </th>
          <th>
            City
          </th>
          <th>
            Telephone
          </th>
          <th
            className="hidden-xs"
          >
            Pets
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a
              href="/owners/1"
            >
              firstName1
               
              lastName1
            </a>
          </td>
          <td
            className="hidden-sm hidden-xs"
          >
            address1
          </td>
          <td>
            city1
          </td>
          <td>
            012345
          </td>
          <td
            className="hidden-xs"
          >
            
          </td>
        </tr>
      </tbody>
    </table>
  </section>
  <a
    className="btn btn-default"
    href="/owners/new"
    onClick={[Function]}
  >
     
    Add Owner
  </a>
</div>
`;
